File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<int> para; 

        for (char i : s){
            if (i == '{'){
                para.push('{'); 
            }else if (i == '('){
                para.push('('); 
            }else if(i == '['){
                para.push('['); 
            }else if (i == '}'){
                if (!para.empty()){
                    if(para.top() == '{'){
                    para.pop(); 
                } else { para.push('}');}
                }else { para.push('}');}
            }else if (i == ')'){
                if (!para.empty()){
                    if(para.top() == '('){
                    para.pop(); 
                } else { para.push(')');} 
                }else { para.push(')');} 
            }else if(i == ']'){
                if (!para.empty()){
                    if(para.top() == '['){
                    para.pop(); 
                } else { para.push(']');}

                }else { para.push(']');} 
            }
        } 
        if (para.empty()){
            return true; 
        }
        return false;   
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        dict = {"(":")", "{":"}", "[":"]"}
        stk = []

        for i in range(len(s)): 
            if s[i] in dict.keys(): 
                stk.append(s[i])
            else: 
                if not stk: 
                    return False
                popped = stk.pop()
                if dict[popped] != s[i]: 
                    return False
        
        return stk == []